src/shm/pool/slot.rs

```diff
@@ -0,0 +1,412 @@
+    #[error("Buffer was already active")]
+    AlreadyActive,
+}
+
+#[derive(Debug)]
+pub struct SlotPool {
+    pub(crate) inner: RawPool,
+    free_list: Arc<Mutex<Vec<(usize, usize)>>>,
+}
+
+#[derive(Debug, Clone)]
+pub struct Slot {
+    inner: Arc<SlotInner>,
+}
+
+#[derive(Debug)]
+struct SlotInner {
+    free_list: Weak<Mutex<Vec<(usize, usize)>>>,
+    offset: usize,
+    len: usize,
+    active_buffers: AtomicUsize,
+}
+
+/// Buffer is counted in active_buffers list; will return to INACTIVE on Release.
+const BUFFER_STATE_ACTIVE: u8 = 0;
+
+/// Buffer is not counted in active_buffers list, but also has not been destroyed.
+const BUFFER_STATE_INACTIVE: u8 = 1;
+
+/// Buffer is counted in active_buffers list; will move to DEAD on Release
+const BUFFER_STATE_DESTROY_ON_RELEASE: u8 = 2;
+
+/// Buffer has been destroyed
+const BUFFER_STATE_DEAD: u8 = 3;
+
+/// Value that is ORed on buffer release to transition to the next state
+const BUFFER_STATE_RELEASE_SET: u8 = 1;
+
+/// Value that is ORed on buffer destroy to transition to the next state
+const BUFFER_STATE_DESTROY_SET: u8 = 2;
+
+#[derive(Debug)]
+pub struct Buffer {
+    buffer: wl_buffer::WlBuffer,
+}
+
+#[derive(Debug)]
+struct BufferData {
+    inner: Arc<SlotInner>,
+    state: AtomicU8,
+}
+
+impl SlotPool {
+    pub(crate) fn new(inner: RawPool) -> Self {
+        let free_list = Arc::new(Mutex::new(vec![(0, inner.len())]));
+        SlotPool { inner, free_list }
+    }
+
+    /// Create a new slot and buffer.
+    ///
+    /// The parameters are:
+    ///
+    /// - `width`: the width of this buffer (in pixels)
+    /// - `height`: the height of this buffer (in pixels)
+    /// - `stride`: distance (in bytes) between the beginning of a row and the next one
+    /// - `format`: the encoding format of the pixels. Using a format that was not
+    ///   advertised to the `wl_shm` global by the server is a protocol error and will
+    ///   terminate your connection
+    ///
+    /// The returned slot may be used to later access the canvas or to create new buffers that
+    /// point to the same data.
+    pub fn create_buffer_slot(
+        &mut self,
+        width: i32,
+        height: i32,
+        stride: i32,
+        format: wl_shm::Format,
+        conn: &Connection, // TODO remove
+    ) -> Result<(Slot, Buffer, &mut [u8]), CreateBufferError> {
+        let len = (height as usize) * (stride as usize);
+        let slot = self.new_slot(len)?;
+        let buffer = self.create_buffer(&slot, width, height, stride, format, conn)?;
+        let canvas = self.raw_data_mut(&slot);
+        Ok((slot, buffer, canvas))
+    }
+
+    /// Get the bytes corresponding to a given slot if drawing to the slot is permitted.
+    ///
+    /// Returns `None` if there are active buffers in the slot or if the slot does not correspond
+    /// to this pool.
+    pub fn canvas(&mut self, slot: &Slot) -> Option<&mut [u8]> {
+        if slot.has_active_buffers() {
```

Comment:
I could see it being useful to access the canvas of one specific buffers in a slot assuming it is not in use.

Reply:
A slot is either used or unused as a whole; it generally only has one buffer.  I do plan to add a `fn canvas(&mut self, buffer: &Buffer)` however.